store-implementation/accumulo-store/src/main/java/uk/gov/gchq/gaffer/accumulostore/MiniAccumuloStore.java

```diff
@@ -0,0 +1,274 @@
+    private static final String BASE_DIRECTORY = "miniAccumuloStoreTest-";
+    private static final String ROOTPW = "rootPW";
+    private MiniAccumuloCluster miniAccumuloCluster = null;
+    private MiniAccumuloConfig miniAccumuloConfig = null;
+    private Connector miniConnector;
+
+    public StoreProperties setUpTestDB(final StoreProperties properties) throws StoreException {
+        setProperties(properties);
+
+        File targetDir = new File("target");
+        File baseDir;
+        if (targetDir.exists() && targetDir.isDirectory()) {
+            baseDir = new File(targetDir, BASE_DIRECTORY + UUID.randomUUID());
+        } else {
+            baseDir = new File(FileUtils.getTempDirectory(), BASE_DIRECTORY + UUID.randomUUID());
+        }
+
+        try {
+            FileUtils.deleteDirectory(baseDir);
+            miniAccumuloConfig = new MiniAccumuloConfig(baseDir, ROOTPW);
+            miniAccumuloConfig.setInstanceName(getProperties().getInstance());
+            miniAccumuloCluster = new MiniAccumuloCluster(miniAccumuloConfig);
+            miniAccumuloCluster.start();
+        } catch (final IOException | InterruptedException e) {
+            throw new StoreException(e.getMessage(), e);
+        }
+
+        // Create the user specified in the properties (if not root)
+        // together with the specified password and give them all authorisations
+        try {
+            if (!getProperties().getUser().equalsIgnoreCase("root")) {
+                miniAccumuloCluster.getConnector("root", ROOTPW).securityOperations()
+                        .createLocalUser(getProperties().getUser(), new PasswordToken(getProperties().getPassword()));
+                miniAccumuloCluster.getConnector("root", ROOTPW).securityOperations()
+                        .grantSystemPermission(getProperties().getUser(), SystemPermission.CREATE_TABLE);
+                miniAccumuloCluster.getConnector("root", ROOTPW).securityOperations()
+                        .grantSystemPermission(getProperties().getUser(), SystemPermission.CREATE_NAMESPACE);
+            }
+            Authorizations auths = new Authorizations("public", "private", "publicVisibility", "privateVisibility", "vis1", "vis2");
+            miniAccumuloCluster.getConnector("root", ROOTPW).securityOperations()
+                    .changeUserAuthorizations(getProperties().getUser(), auths);
+        } catch (final AccumuloException | AccumuloSecurityException e) {
+            throw new StoreException(e.getMessage(), e);
+        }
+
+        // Create the new properties object to pass back, including connection items
+        AccumuloProperties accumuloProperties = (AccumuloProperties) properties.clone();
+        accumuloProperties.setInstance(miniAccumuloCluster.getInstanceName());
+        accumuloProperties.setZookeepers(miniAccumuloCluster.getZooKeepers());
+
+        return accumuloProperties;
+    }
+
+    public void tearDownTestDB() {
+        this.closeMiniAccumuloStore();
```

Comment:
What would be the impact if the store was not closed down? I only ask because if we could remove this method, we wouldn't have to pollute the Store class with code only designed to be used with tests.

Reply:
If there was no handling for disposal of the Mini Accumulo instance then the processes may continue to run, and the filesystem space used by the instance will not be recovered, gradually building up over time in the target directory.
Adding the 2 concrete methods to the store class was not done lightly, was asked about early on and was accepted as a necessary evil.
Adding them to the store class means that the mechanism for a store that requires creation and destruction of the store for test purposes can be used for all other store types. Indeed, it is looking like federated-store may need to make use of the mechanism.
They were made as compact and efficient as possible.
Making them concrete methods that are then overridden where needed rather than abstract methods means the other Store subclasses do not need to be changed to have what will simply be a copy of what has been put in the Store class.
As regards polluting the main code with functionality only used in tests I agree it is not ideal, but whether it ended up in Store, AccumuloStore or MiniAccumuloStore (it being derived from the deprecated MockAccumuloStore) makes no difference from that point of view - all are in the main code base.